### PR TITLE
Breaks Formatters up until individual Formatter classes, propagates flags/state

### PR DESCRIPTION
**Pull request by HaloFour**
Breaks up the `Formatters` class into multiple `Formatter` implementations with a `FormatterChain` that, similarly to `HandlerChain`, can take a `Seq[Formatter]` and apply each of them to read/write HTTP headers for the given SpanID.   This could be a first pass at making the formatters configurable.

Reuses the OpenTelemetry `HttpTraceContext` implementation as a `Formatter` to read/write the `traceparent` and `tracestate` HTTP headers.

### CODE REVIEW

**Comment by HaloFour** [comment]
Apologies for the rapid-fire PRs and the accumulation of commits into those PRs.  I've adopted this project as my Labweek project this week so I've been able to dedicate a lot of time towards it.  For this PR I can certainly rollback the commit that splits things out and makes it configurable and submit that as a separate PR.

I'm also on the fence as to whether `Formatter` should be moved to `money-api` and made into a more Java-friendly interface.

File: money-core/src/main/scala/com/comcast/money/core/formatters/FormatterChain.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.comcast.money.core.formatters
+import com.comcast.money.api.SpanId
+import com.typesafe.config.Config
+import scala.collection.JavaConverters._
+
+case class FormatterChain(formatters: Seq[Formatter]) extends Formatter {
```
Comment (ptravers) [inline money-core/src/main/scala/com/comcast/money/core/formatters/FormatterChain.scala:22]:
```suggestion
final case class FormatterChain(formatters: Seq[Formatter]) extends Formatter {
```

File: money-core/src/main/scala/com/comcast/money/core/formatters/FormatterChain.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    formatter => formatter.fromHttpHeaders(getHeader, log)
+  } headOption
+
+  override def fields: Seq[String] = formatters.flatMap {
+    formatter => formatter.fields
+  }
+
+  override def setResponseHeaders(getHeader: String => String, addHeader: (String, String) => Unit): Unit = formatters.foreach {
+    formatter => formatter.setResponseHeaders(getHeader, addHeader)
+  }
```
Comment (ptravers) [inline money-core/src/main/scala/com/comcast/money/core/formatters/FormatterChain.scala:37]: do we want to offer some information as to the total size of headers we are adding?

File: money-core/src/main/scala/com/comcast/money/core/formatters/MoneyTraceFormatter.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import io.opentelemetry.trace.{ TraceFlags, TraceState }
+
+import scala.util.{ Failure, Success, Try }
+
+object MoneyTraceFormatter {
+  private[core] val MoneyTraceHeader = "X-MoneyTrace"
+  private[core] val MoneyHeaderFormat = "trace-id=%s;parent-id=%s;span-id=%s"
+}
+
+class MoneyTraceFormatter extends Formatter {
```
Comment (ptravers) [inline money-core/src/main/scala/com/comcast/money/core/formatters/MoneyTraceFormatter.scala:29]:
```suggestion
final class MoneyTraceFormatter extends Formatter {
```

File: money-core/src/main/scala/com/comcast/money/core/formatters/OtelFormatter.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    override def addEvent(name: String, attributes: Attributes): Unit = ()
+    override def addEvent(name: String, attributes: Attributes, timestamp: Long): Unit = ()
+    override def setStatus(canonicalCode: StatusCanonicalCode): Unit = ()
+    override def setStatus(canonicalCode: StatusCanonicalCode, description: String): Unit = ()
+    override def recordException(exception: Throwable): Unit = ()
+    override def recordException(exception: Throwable, additionalAttributes: Attributes): Unit = ()
+    override def updateName(name: String): Unit = ()
+    override def `end`(): Unit = ()
+    override def `end`(endOptions: EndSpanOptions): Unit = ()
+    override def isRecording: Boolean = false
```
Comment (ptravers) [inline money-core/src/main/scala/com/comcast/money/core/formatters/OtelFormatter.scala:65]: will this cause any issues in the propagator? we should probably create an integration test with otel and zipkin

**Review by ptravers — APPROVED** [review]

File: money-core/src/main/scala/com/comcast/money/core/formatters/FormatterChain.scala (hunk 2 above)
Comment (HaloFour) [reply money-core/src/main/scala/com/comcast/money/core/formatters/FormatterChain.scala:37]: The `fields` member would return the total list of headers that might be added.  The actual headers depend on the SpanId/SpanContext.

Comment (ptravers) [reply money-core/src/main/scala/com/comcast/money/core/formatters/FormatterChain.scala:37]: I meant more that we might make the request breach max header size for lets say apache. wasnt sure if maybe it was worth giving the user some information about the size in bytes that we are adding

File: money-core/src/main/scala/com/comcast/money/core/formatters/OtelFormatter.scala (hunk 4 above)
Comment (HaloFour) [reply money-core/src/main/scala/com/comcast/money/core/formatters/OtelFormatter.scala:65]: I don't believe so.  All of the OTel implementations grab the `SpanContext` and work with that.  There really isn't any useful information that the propagator could glean from the full `Span` API, the only member that returns a value is `isRecording`.

File: money-core/src/main/scala/com/comcast/money/core/formatters/FormatterChain.scala (hunk 2 above)
Comment (HaloFour) [reply money-core/src/main/scala/com/comcast/money/core/formatters/FormatterChain.scala:37]: I don't think that would be a concern.  Even if someone turned on every single propagator I've included I don't think the total size of the headers + values would exceed a few hundred bytes.

**Pull request merged**